include/mrdox/Support/SharedPtr.hpp

```diff
@@ -0,0 +1,325 @@
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// Copyright (c) 2023 Vinnie Falco (vinnie.falco@gmail.com)
+//
+// Official repository: https://github.com/cppalliance/mrdox
+//
+
+#ifndef MRDOX_API_SUPPORT_SHAREDPTR_HPP
+#define MRDOX_API_SUPPORT_SHAREDPTR_HPP
+
+#include <mrdox/Platform.hpp>
+#include <atomic>
+
+namespace clang {
+namespace mrdox {
+
+template<class T>
+class SharedPtr;
+
+template<class T>
+class AtomicSharedPtr;
+
+template<class U, class... Args>
+SharedPtr<U>
+makeShared(Args&&...);
+
+//------------------------------------------------
+
+class SharedPtrBase
+{
+protected:
+    struct Base
+    {
+        virtual ~Base() = default;
+        Base() noexcept = default;
+        Base* addref() noexcept;
+        void release() noexcept;
+        virtual void* get() noexcept = 0;
```

Comment:
I'm wondering if this needs to be a virtual function. Could you use a templated `Base` instead?

Reply:
If `Base` was a class template then it would be undefined behavior to cast `Base<T>*` to `Base<U>*`, which would be needed for those functions.

Reply:
Ah right, inheritance